src/main/scala/com/courseWork/auth/AuthService.scala

```diff
@@ -0,0 +1,103 @@
+package com.courseWork.auth
+
+import cats.syntax.apply._
+import cats.syntax.applicative._
+import cats.syntax.flatMap._
+import cats.{Applicative, FlatMap, Functor}
+import cats.data.{Validated, ValidatedNec}
+
+import java.util.UUID
+
+trait AuthService[F[_]] {
+  def register(nickname: String, email: String, password: String): F[User]
+  def login(data: LoginData): F[Token]
+}
+
+class AuthServiceImpl[F[_]: FlatMap: Applicative: Functor](userRepository: UserRepository[F]) extends AuthService[F] {
+  type UserValidationErrorsOr[A] = ValidatedNec[UserValidationError, A]
+
+  override def register(nickname: String, email: String, password: String): F[User] = {
+    validateAndGet(nickname, email, password) match {
+      case Validated.Valid(user)         => userRepository.add(user)
+      case Validated.Invalid(errorChain) => ??? // ae.raiseError() // todo: Make MyOwnErrors
+    }
+  }
+
+  override def login(data: LoginData): F[Token] = Token("tokentokentokentokentokentokentokentoken").pure[F]
+
+  private def validateNickname(nickname: String): UserValidationErrorsOr[String] =
```

Comment:
It is better to use domain models for such purposes. In Scala 3 u can use 
```scala
opaque type Nickname
object Nickname {
  def apply(nick: String): Nickname = ??? // some validation here
}
```

In Scala 2.13:
```scala
case class Nickname(value: String) extends AnyVal
object Nickname {
  // the same stuff here as above
}
```
For better performance (if needed) - opensource lib: newtype (just google newtype Scala)